apps/backend/api/models/ContextWidget.js

```diff
@@ -0,0 +1,314 @@
+          SET "order" = CASE id 
+            ${reorderedWidgets.map(w => `WHEN ${w.id} THEN ${w.order}`).join('\n')}
+          END
+          WHERE id IN (${reorderedWidgets.map(w => w.id).join(',')})
+        `
+        await bookshelf.knex.raw(query).transacting(trx)
+      }
+
+      // Set the removed widget's order and parent_id to null
+      await removedWidget.save({ order: null }, { 
+        patch: true,
+        parent_id: null,
+        transacting: trx 
+      })
+
+      return removedWidget
+    }
+
+    if (existingTrx) {
+      return doWork(existingTrx)
+    }
+
+    return await bookshelf.transaction(trx => doWork(trx))
+  },
+  reorder: async function({ id, order, trx: existingTrx }) {
+    const doWork = async (trx) => {
+      const movedContextWidget = await ContextWidget.where({ id }).fetch({ transacting: trx })
+      if (!movedContextWidget) throw new Error('Context widget not found')
+        
+      const peerWidgets = await movedContextWidget.peers(trx).fetch({ transacting: trx })
+      const addItToTheEnd = peerWidgets.find(w => w.get('id') === id) ? peerWidgets.length : peerWidgets.length + 1
+
+      const newOrder = order || addItToTheEnd
+      // Check if this is a root widget trying to move to the home position
+      if (!movedContextWidget.get('parent_id') && newOrder === 1) {
+        throw new Error('The home widget must remain the first widget in the context menu')
+      }
+    
+      // Find old order if widget exists in peers
+      const oldOrder = peerWidgets.find(w => w.get('id') === id)?.get('order')
+    
+      // Get widgets that need order updates
+      const reorderedWidgets = peerWidgets.reduce((acc, widget) => {
+        const currentOrder = widget.get('order')
+        const widgetId = widget.get('id')
+        // Skip widgets outside the affected range
+        if (oldOrder) {
+          if (currentOrder < Math.min(oldOrder, newOrder) || 
+              currentOrder > Math.max(oldOrder, newOrder)) {
+            return acc
+          }
+        } else if (currentOrder < newOrder) {
+          return acc
+        }
+    
+        // Calculate new order
+        let newWidgetOrder = currentOrder
+        if (currentOrder >= newOrder) newWidgetOrder = currentOrder + 1
+        if (oldOrder && currentOrder > oldOrder) newWidgetOrder = currentOrder - 1
+        if (widgetId === id) newWidgetOrder = newOrder
+        if (newWidgetOrder !== currentOrder) acc.push({ id: widgetId, order: newWidgetOrder })
+
+        return acc
+      }, [])
+
+      // If widget doesn't have an order yet, add it to reorderedWidgets
+      if (movedContextWidget.get('order') === null) {
+        reorderedWidgets.push({ id, order: newOrder })
+      }
+
+      // Update all affected widgets in a single query
+      const query = `
+        UPDATE context_widgets 
+        SET 
+          "order" = CASE id 
+            ${reorderedWidgets.map(w => `WHEN ${w.id} THEN ${w.order}`).join('\n')}
```

Comment:
check out the reorderPostInCollection collection mutation to see I do it. I think my way would be slightly more efficient (definitely less code). but maybe not. no big deal

Reply:
`reorderPostInCollection` handles reordering on a single set. Reorder tree handles reordering on two levels; the parent and child levels, and moves in between. The tests might help illustrate this.